pio/uart_pio_dma/uart_pio_dma.c

```diff
@@ -0,0 +1,274 @@
+/**
+ * Copyright (c) 2024 Raspberry Pi (Trading) Ltd.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <stdio.h>
+#include <string.h>
+
+#include "pico/stdlib.h"
+#include "hardware/pio.h"
+#include "hardware/uart.h"
+#include "hardware/dma.h"
+#include "uart_rx.pio.h"
+#include "uart_tx.pio.h"
+
+// ------- USER CONFIGURATION -----
+
+// Send data via GPIO 4 and receives it on GPIO 5
+// *** You must conect these pins with a wire ***
+#define GPIO_TX 4
+#define GPIO_RX 5
+
+// Use PIO instead of real UART for receiving data
+#define USE_PIO_FOR_RX 1
+
+// Use DMA rather than polling when receiving data
+#define USE_DMA_FOR_RX 1
+
+// Use PIO instead of real UART for sending data
+#define USE_PIO_FOR_TX 1
+
+// Use DMA rather than polling when sending data
+#define USE_DMA_FOR_TX 1
+
+#define SERIAL_BAUD 921600
+#define HARD_UART_INST uart1
```

Comment:
Out of curiosity I just did some quick local compile-tests, and this code all still compiles if you replace
```c
#define HARD_UART_INST uart1
```
with
```c
#if !USE_PIO_FOR_RX || !USE_PIO_FOR_TX
#define HARD_UART_INST uart1
#endif
```
:wink:  (which _perhaps_ makes it clearer that the "hard UART" isn't used at all when PIO is being used for both RX and TX?)